src/apack/zcl_abapgit_apack_migration.clas.abap

```diff
@@ -0,0 +1,217 @@
+          RECEIVING
+            result = lo_factory.
+
+        CALL METHOD lo_factory->('CREATE_CLIF_SOURCE')
+          EXPORTING
+            clif_name = is_clskey-clsname
+          RECEIVING
+            result    = lo_source.
+
+        TRY.
+            CALL METHOD lo_source->('IF_OO_CLIF_SOURCE~LOCK').
+          CATCH cx_oo_access_permission.
+            zcx_abapgit_exception=>raise( 'source_new, access permission exception' ) ##NO_TEXT.
+        ENDTRY.
+
+        lt_source_code = get_interface_source( ).
+
+        CALL METHOD lo_source->('IF_OO_CLIF_SOURCE~SET_SOURCE')
+          EXPORTING
+            source = lt_source_code.
+
+        CALL METHOD lo_source->('IF_OO_CLIF_SOURCE~SAVE').
+        CALL METHOD lo_source->('IF_OO_CLIF_SOURCE~UNLOCK').
+
+      CATCH cx_sy_dyn_call_error.
+        add_interface_source_classic( is_clskey = is_clskey ).
+    ENDTRY.
+
+  ENDMETHOD.
+
+
+  METHOD run.
+
+    DATA: lo_apack_migration TYPE REF TO zcl_abapgit_apack_migration.
+
+    CREATE OBJECT lo_apack_migration.
+    lo_apack_migration->perform_migration( ).
+
+  ENDMETHOD.
+
+
+  METHOD create_interface.
+
+    DATA: ls_interface_properties TYPE vseointerf,
+          ls_clskey               TYPE seoclskey,
+          ls_inactive_object      TYPE dwinactiv,
+          lt_inactive_objects     TYPE TABLE OF dwinactiv.
+
+    ls_interface_properties-clsname  = c_interface_name.
+    ls_interface_properties-version  = '1'.
+    ls_interface_properties-langu    = 'E'.
+    ls_interface_properties-descript = 'APACK: Manifest interface' ##NO_TEXT.
+    ls_interface_properties-exposure = '2'.
+    ls_interface_properties-state    = '1'.
+    ls_interface_properties-unicode  = abap_true.
+
+    CALL FUNCTION 'SEO_INTERFACE_CREATE_COMPLETE'
+      EXPORTING
+        devclass        = '$TMP'
+      CHANGING
+        interface       = ls_interface_properties
+      EXCEPTIONS
+        existing        = 1
+        is_class        = 2
+        db_error        = 3
+        component_error = 4
+        no_access       = 5
+        other           = 6
+        OTHERS          = 7.
+    IF sy-subrc <> 0.
+      zcx_abapgit_exception=>raise( 'Error from SEO_INTERFACE_CREATE_COMPLETE' ) ##NO_TEXT.
+    ENDIF.
+
+    ls_clskey-clsname = c_interface_name.
+
+    add_interface_source( is_clskey = ls_clskey ).
+
+    ls_inactive_object-object   = 'INTF'.
+    ls_inactive_object-obj_name = c_interface_name.
+    INSERT ls_inactive_object INTO TABLE lt_inactive_objects.
+
+    CALL FUNCTION 'RS_WORKING_OBJECTS_ACTIVATE'
+      TABLES
+        objects                = lt_inactive_objects
+      EXCEPTIONS
+        excecution_error       = 1
+        cancelled              = 2
+        insert_into_corr_error = 3
+        OTHERS                 = 4.
+    IF sy-subrc <> 0.
+      zcx_abapgit_exception=>raise( 'error from RS_WORKING_OBJECTS_ACTIVATE' ) ##NO_TEXT.
+    ENDIF.
+
+  ENDMETHOD.
+
+
+  METHOD get_interface_source.
+
```

Comment:
Can you insert also some comment that the interface was automatically generated and can be removed at any point of time without any harm to abapgit?

Reply:
Well, if it's removed, it is re-created again after each start of abapGit. So such a comment doesn't make much sense IMHO. However, as with any other automatically created artifact, it can be safely removed if abapGit needs to be uninstalled - instructions are provided on the respective page.

Reply:
The auto-recreation feature convinces me even more that the interface needs some kind of a mark that it was created by abapgit and will be created until the user stop using abapgit. Any time I talk to my support colleagues working on customer systems about abapgit, they are afraid of side effects - this iface is one of the side effects and must be clearly documented IMHO.

Reply:
A possible alternative - use only the manifest file? I am not sure why we need the class, the interface and the xml file while one would expect only the xml file. I apologize for my ignorance.

Reply:
the interface implementation would go into the target project source, along with the manifest file, this can be seen as duplicated information. Personally, I'd like the ABAP code folder /src/ to only contain code that is valid for the project, I would expect the interface to be flagged by static analysis.

Alternative approaches:
A: Add all information in the `.abapgit.xml` file, it already exists and has an editor, no ABAP artifacts persisted in system
B: Build an editor that is independent of the abapGit source tree, and add an exit in abapGit which then discovers the extra file
C: Put the interface plus APACK specific code in a separate project, when APACK is installed the user can decide to put it to transportable package.

Reply:
@jfilak Well, if we only had the file itself, we'd need a possibility to edit it - and I intentionally wanted the information to be available easily both in the system (as code) and in the Git repository. Future enhancements could (no promises made, of course) use the same interface implementation to automatically retrieve dependencies independently from abapGit from a central repository like https://mvnrepository.com/. That's also why I didn't include it directly in the abapGit persistence, @larshp

Reply:
And one more thing, @jfilak: To my knowledge there is no easy way to store a simple file like such an XML on the ABAP server in an ABAP package - that's why "configuration-as-code" is done via an interface implementation.

Reply:
Would it be possible to use only Dynamic ABAP programming without the interface? I.e. you could search for a class with name matching the pattern 'z{tdevc-name}apack', create its instance dynamically and get the configuration in the form of an internal table with lines of a character type. There could be a standard implementation reading the apack manifest from .abapgit.xml for abapGit projects but it would also be possible to create own random class without the need to "mess" your system with abapgit. Anyways, I am not blocking this PR, I am just trying to oppose to find the best solution.

Reply:
Well, sometimes the "best solution" is certainly a matter of taste or personal preference - or the choice of one compromise over the other, right? ;) If you ask me, I wouldn't go for a completely dynamic solution. Possible error sources are endless when people use the wrong type, have slightly changed names, don't update the structure/table types if there are new manifest elements in future versions etc. Of course, it's technically feasible, but I wouldn't recommend it...

I also understand the arguments of many customers that they are afraid of deploying software like abapGit to their systems and are concerned about additional objects in their system - but in this case it's only an interface with no strings attached. Virtually any other project - especially in the Cloud - has a bigger footprint.

But coming back to APACK in general: @larshp and me will probably have a in-depth discussion how to move on in that area as he will visit the HQ next week. From my current point of view (under the conditions I mentioned before in this thread) the only viable alternative is the aforementioned option C - to move APACK to a dedicated project. By this, also the marker interface would come as dedicated object and the only thing which needs to be done is to include everything into a transportable package (OK, also error-prone, but probably less than now). abapGit would only be "tainted" by an additional user exit and users are completely free to decide if they deploy it or not.

However, there are also several downsides of this approach, which I don't want to hide: 

1. **Adoption:** If it's a separate project, people need to deploy and update it independently from abapGit which would IMHO severely endanger the probability of adoption in the - anyway small - community. 
2. **User Experience and Maintainability:** At some point in time you want to go beyond pure serialization - the fun starts when dependencies are automatically resolved during the import of software. However, building it separately virtually creates the need for many parallel structures to bring in ABAP code into the system. abapGit is usually deployed as single report which makes it more or less impossible to re-use functionality from other components. Therefore, a separate APACK solution in the backend would create the need for an own UI and the whole code behind the curtain. Well, I have doubts that somebody would maintain that - or that users would accept two parallel UIs. That makes it unlikely that such a separate solution would use abapGit as one possible foundation and rather go _exclusively_ for other repository types and be Cloud-only such as the technology used here: https://developers.sap.com/tutorials/abap-environment-gcts.html As said - I'd rather keep it agnostic of a concrete serialization/repository type - in the end it's just a format definition that can be adopted by whoever wants it. ;)
3. **Chances that it becomes real:** Putting it into a separate repository makes it an independent project. I can't speak for the whole company of course, but with the focus on the Cloud, it's not that likely that there will be approvals for a completely separate open source project with a 100% on-premise focus. We only have approvals for upstream commits to abapGit (and related things such as the ADT plugin) so far. So, in the end somebody else would need to implement and maintain it - and we've seen how far we went since #217, unfortunately. :( Resources are limited and we should make the best out of what we have...

Reply:
>  'z{tdevc-name}apack'

This won't work for code in namespaces, which have eg /acme/ instead of the z in front of it.
But this seems to me to work better than having zif_apack_manifest AND if_apack_manifest with one beeing only available where the other is.